### PR TITLE
Add const generics to infer (and transitive dependencies)

### PR DESCRIPTION
**Pull request by varkor**
Split out from #53645. This work is a collaborative effort with @yodaldevoid.

There are a number of stubs. These are mainly to ensure we don't overlook them when completing the implementation, but are not necessary for the initial implementation. We plan to address these in follow up PRs.

r? @eddyb / @nikomatsakis

### CODE REVIEW

File: src/librustc/ich/impls_ty.rs
Comment (eddyb) [inline src/librustc/ich/impls_ty.rs]: You likely need `PlaceholderConst` too.

**Review by eddyb — CHANGES_REQUESTED** [review]
LGTM, modulo comments, some of which require someone who understands certain typesystem interactions - I've pinged @nikomatsakis but I suspect @rust-lang/wg-traits can also help here.

File: src/librustc/infer/canonical/mod.rs
Comment (varkor) [reply src/librustc/infer/canonical/mod.rs]: Rebasing is such a pain.

File: src/librustc/infer/const_variable.rs
Comment (varkor) [reply src/librustc/infer/const_variable.rs]: We had something more basic initially, but kept running into situations where an analogous abstraction to `TypeVariableTable` didn't exist. Replicating a similar data structure for consts made everything clearer, but it's possible that some of what we were trying to do is simply unnecessary at that level of generality for consts.

> The problem with types is that there are "integer infer variables" and "float infer variables", but constants only have one kind of inference variable.

Is this not a separate concern (i.e. there's `type_variables`, `int_unification_table` and `float_unification_table` and at the moment, `const_variables` follows the structure laid out by `type_variables` rather than the other two)? It might be easiest to identify places where the const unification table is doing more work than necessary, so we can work out if it isn't actually necessary.

Comment (eddyb) [reply src/librustc/infer/const_variable.rs]: Oh I just thought there is already something like this, then I don't know.

Is there no `..._table` field that has a type that's... not from librustc? Some general-purposes "unification table"?

File: src/librustc/infer/canonical/mod.rs
Comment (eddyb) [reply src/librustc/infer/canonical/mod.rs]: Sure, but this PR makes very little sense atm anywhere "lazy" shows up. We'd have to fix it either before or after, and I want to be able to review this PR instead of just skimming it.

File: src/librustc/infer/const_variable.rs
Comment (varkor) [reply src/librustc/infer/const_variable.rs]: I don't see one in `InferCtxt` at least. It could make sense to create an abstraction if it doesn't exist (and I'm not just overlooking it somewhere) as a `C-cleanup` issue afterwards.

Comment (eddyb) [reply src/librustc/infer/const_variable.rs]: I just checked and it turns out that I was referring to `int_unification_table` and `float_unification_table`, which rely on e.g.:
https://github.com/rust-lang/rust/blob/aa97448caf49ee25271a5ad58482283404b9144f/src/librustc/infer/unify_key.rs#L8-L13

The reason this wasn't immediately obvious was that you have a `values` field and that it *doesn't* hold the actual variable values (maybe it used to, for `TypeVariableTable`, at some point?) *but it's really* "origins" (and should be named as such; in `TypeVariableTable` too).

Note that the `UnificationTable` holds `<Vid as UnifyKey>::Value` for each active vid, and you could combine `ConstVariableData` and `ConstVariableValue` into one type to use as that `Value` associated type.

A bit surprised to see integer/float inference variables don't have an origin attached, but they could all use the same mechanism.

IMO the reason `type_variables` exists as a separate module is mainly because of subtyping relations, and we shouldn't need that here.

File: src/librustc/infer/const_variable.rs
Comment (eddyb) [inline src/librustc/infer/const_variable.rs]: I just realized all of these have `Span`s, maybe that can be factored out? (but this is probably not the only place)

File: src/librustc/infer/const_variable.rs
Comment (eddyb) [inline src/librustc/infer/const_variable.rs]: @nikomatsakis I'm thinking maybe there should be a generic `enum` like this to share between types and consts.

File: src/librustc/infer/const_variable.rs
Comment (eddyb) [inline src/librustc/infer/const_variable.rs]: Is this enough? Don't we also need the type?

File: src/librustc/infer/const_variable.rs
Comment (eddyb) [inline src/librustc/infer/const_variable.rs]: Yeah, this entire `impl` is just universal boilerplate.

File: src/librustc/infer/canonical/mod.rs
Comment (oli-obk) [reply src/librustc/infer/canonical/mod.rs]: Ok, I'll create a PR cleaning up the situation

File: src/librustc/infer/canonical/mod.rs
Comment (eddyb) [inline src/librustc/infer/canonical/mod.rs]: Sounds like `PlaceholderConst` should contain the type.

Comment (varkor) [reply src/librustc/infer/canonical/mod.rs]: Yeah, it probably should. The reason I didn't do it earlier was it ended up propagating `<'tcx>` everywhere and I wanted to make sure it was definitely necessary before going down that route.

Comment (yodaldevoid) [reply src/librustc/infer/canonical/mod.rs]: I attempted propagating `<'tcx>` around when originally adding `CanonicalVarKind::Const` and found it to be impossible because `CanonicalVarInfo`s are allocated in the global arena.

Comment (eddyb) [reply src/librustc/infer/canonical/mod.rs]: Sound like we need to bother @nikomatsakis.

But also, aren't const types supposed to be always be "global"?

Comment (yodaldevoid) [reply src/librustc/infer/canonical/mod.rs]: AFAIK, types are only global after they have been lifted and their dependencies on type parameters and such are taken care of.

Comment (eddyb) [reply src/librustc/infer/canonical/mod.rs]: Type parameters are "global" in this sense, only inference variables aren't.
I'm saying that I don't think we should ever have a constant value in the type system with an *uninferred type*.

File: src/librustc/infer/canonical/mod.rs
Comment (crlf0710) [reply src/librustc/infer/canonical/mod.rs]: #59178 is merged.

**Comment by varkor** [comment]
This is currently blocked on https://github.com/rust-lang-nursery/ena/pull/19 and response from @nikomatsakis regarding the review comments above.

**Comment by crlf0710** [comment]
ena 0.12.0 is [released](https://docs.rs/ena/0.12.0/ena/).

**Comment by varkor** [comment]
I missed a detail in the previous pull request to ena: this is now blocked on https://github.com/rust-lang-nursery/ena/pull/21.

File: src/librustc/infer/const_variable.rs
Comment (varkor) [reply src/librustc/infer/const_variable.rs]: I'll refactor `TypeVariableOrigin` and `ConstVariableOrigin` in a follow up.

Comment (varkor) [reply src/librustc/infer/const_variable.rs]: Opened https://github.com/rust-lang/rust/pull/59331.

**Comment by eddyb** [comment]
@varkor This should be unblocked now, just released `ena v0.13`.

**Comment by varkor** [comment]
This is now blocked on https://github.com/rust-lang/rust/pull/59415.

**Comment by varkor** [comment]
I've addressed all the comments that aren't waiting on feedback from @nikomatsakis.

**Comment by Centril** [comment]
Reassigning the remainder of the review to r? @nikomatsakis then.

**Comment by eddyb** [comment]
Anyone else from @rust-lang/wg-traits want to take a look at the questions I left for @nikomatsakis, too?

**Comment by WiSaGaN** [comment]
Does this still block on @nikomatsakis and @rust-lang/wg-traits?

**Comment by crlf0710** [comment]
@WiSaGaN Yes, althrough niko has responded that he'll add it to his list last week:
https://rust-lang.zulipchat.com/#narrow/stream/144729-wg-traits/topic/.2359008.20changes.20to.20infer

**Comment by nikomatsakis** [comment]
My great apologies for the delay. I've scheduled this for review this coming Wednesday at 12:00 UTC-04:00.

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
OK, [as I noted on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/const.20generics.20in.20.60infer.60.20.2359008/near/163581003), I'm going to need to spend more time reviewing. I've tentatively scheduled the next review session for April 22 at 15:00 UTC-04:00. 

Apart from the random nits I accumulated below, the main thing I noticed while going through this PR is that it doesn't have any tests. To that end, I asked @varkor if they could bring in some tests -- in part because I intend to walk through some of the tests to get a feeling for how the system overall is fitting together. 

I didn't yet try to address @eddyb's various questions =) but I'll come back to those once I have a better feeling for the overall picture here.

File: src/librustc/ty/relate.rs
Comment (varkor) [reply src/librustc/ty/relate.rs]: Ah, that does seem plausible. I'll remove this assertion.

File: src/librustc/infer/const_variable.rs
Comment (varkor) [reply src/librustc/infer/const_variable.rs]: Not as far as I know. This was probably an oversight.

File: src/librustc/infer/canonical/canonicalizer.rs
Comment (varkor) [reply src/librustc/infer/canonical/canonicalizer.rs]: I think this code has been refactored since.

**Comment by nikomatsakis** [comment]
OK, unfortunately I got too busy today and failed to do the review I had planned to do. I've rescheduled that to Tuesday (tomorrow) at 16:00 UTC-04:00.

File: src/librustc/infer/combine.rs
Comment (nikomatsakis) [reply src/librustc/infer/combine.rs]: Well, the other variants of `InferConst` are:

```rust
/// An inference variable for a const, for use in const generics.
#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd,
         Ord, RustcEncodable, RustcDecodable, Hash, HashStable)]
pub enum InferConst<'tcx> {
    /// Infer the value of the const.
    Var(ConstVid<'tcx>),
    /// A fresh const variable. See `infer::freshen` for more details.
    Fresh(u32),
    /// Canonicalized const variable, used only when preparing a trait query.
    Canonical(DebruijnIndex, BoundVar),
}
```

so, yes, this is due to the placeholder stuff. I don't know whether it will get "cleaned up" or not, I guess that depends a bit on what the final shape of our type story looks like, which is something I've been thinking a lot about (I'd like to share the code between rustc and chalk, for example). 

I think overall I'd like to move away from the "zoo" of variants that exist today and have three:

- free variables: inference variables (existentially quantified) and placeholders (universally quantified)
- bound variables (based on debruijn indices): used for canonicalization, `for<'a>` types, etc.

but that's probably a larger conversation than just this PR.

Comment (nikomatsakis) [reply src/librustc/infer/combine.rs]: I think in short this part of the code seems ok to land as is, it seems comparable to the other cases, and we can refactor them together.

File: src/librustc/infer/mod.rs
Comment (nikomatsakis) [reply src/librustc/infer/mod.rs]: It seems like it could work on any type-foldable.

Comment (nikomatsakis) [reply src/librustc/infer/mod.rs]: It's not something you would normally want to do on a composite structure, but I don't see the *harm* in doing so -- presumably the only actual callers would be people giving in a `Ty` or `Const` directly.

Comment (nikomatsakis) [reply src/librustc/infer/mod.rs]: So in summary: yes, I think we should make it generic, rather than introduce multiple variants with different names.

File: src/librustc/ty/relate.rs
Comment (nikomatsakis) [reply src/librustc/ty/relate.rs]: Well, to be quite honest, I'm a bit confused by what the "normalization policy" is here. If this were doing eager normalization, as I would expect at the moment, then this doesn't seem wrong. But if it's lazy normalization, then indeed it is wrong. It seems like we can land it as is, but I think that overall it would be good to try and document and talk out the design of constants and const generics.

Comment (nikomatsakis) [reply src/librustc/ty/relate.rs]: Design meeting! :)

File: src/librustc/infer/const_variable.rs
Comment (nikomatsakis) [reply src/librustc/infer/const_variable.rs]: Yeah, so, I rewrote ena for lark and I made a bunch of changes. As I recall, one of the things I changed was eliminating this sort of boilerplate (I had some other goals too, like being able to track the tree of unifications for better error messages). To be honest, I sort of *forget* what I did in Lark. I think this is another work item that is worth trying to postpone till after the PR lands, and then cleanup separately.

File: src/librustc/infer/const_variable.rs
Comment (nikomatsakis) [reply src/librustc/infer/const_variable.rs]: Yep. As above, I think we ought to land as is, but we should definitely make some time to talk about how to do this better.

File: src/librustc/infer/const_variable.rs
Comment (nikomatsakis) [reply src/librustc/infer/const_variable.rs]: Well, I think the type is carried in the `ty::Const` wrapper, for better or worse:

```rust
    #[inline]
    pub fn mk_const_var(self, v: ConstVid<'tcx>, ty: Ty<'tcx>) -> &'tcx Const<'tcx> {
        self.mk_const(ty::Const {
            val: ConstValue::Infer(InferConst::Var(v)),
            ty,
        })
    }
```

File: src/librustc/infer/canonical/mod.rs
Comment (nikomatsakis) [reply src/librustc/infer/canonical/mod.rs]: Hmm. Something feels wrong here indeed. I'm trying too put my finger on what it is. I guess in short I don't like the idea of there being a **type** in the canonical-var-info -- I feel like types should **in the canonicalized value**.

I think what I would expect is that the placeholder doesn't replace the entire `ty::Const`, but rather **just** the `ConstValue`. In this way, it can be instantiated without needing the type. 

(Similarly, in the existential case that appears above, you wouldn't make a "Fresh type variable" to represent its type -- that also seems wrong.)

File: src/librustc/infer/const_variable.rs
Comment (varkor) [reply src/librustc/infer/const_variable.rs]: It's not carried around in `Unknown`, which I think is the part @eddyb was referring to. (There's some other discussion about this elsewhere, but there's a `FIXME` in the code about this now.)

**Comment by nikomatsakis** [comment]
OK, I think I answered all of the outstanding comments from @eddyb. I've read through enough of this PR that I feel comfortable with it landing overall -- but I think I would really like to have some time to talk out the overall strategy, especially around normalization. 

@varkor and @yodaldevoid -- maybe we can sync up a bit next week and you can try to talk me through it? (Over Zulip would be fine.) I'm interested in having the design here [written out and submitted as a proposal](https://github.com/rust-lang/compiler-team/pull/75). Anyway, I'll ping you on Zulip to discuss.

As far as the concerns @eddyb raised, I think they fall into two categories:

- Boilerplate around inference variables: I think we should table this concern but revisit it in a uniform way outside of this PR.
- The integration of const values into canonicalization: [Something does indeed seem wrong here](https://github.com/rust-lang/rust/pull/59008#discussion_r279094359), though I'm not sure if it should block this PR per s, as I guess none of this is "exposed" to end useres.

**Comment by varkor** [comment]
I've fixed the remaining issues regarding refactoring (that are in-scope for this PR) and rebased. As far as I can tell now, there's just the question of whether we want to settle the questions about canonicalisation before we merge this PR, or whether we can leave it for afterwards?

**Comment by eddyb** [comment]
Broken canonicalization will probably prevent MIR borrowck from working, but we might not need that yet.

r=me unless @nikomatsakis disagrees

**Comment by nikomatsakis** [comment]
I think we can settle this stuff afterwards and am fine with merging

**Comment by nikomatsakis** [comment]
Well, wait, will this potentially affect things like `[u8; 22]`? I'm assuming not

**Comment by varkor** [comment]
It shouldn't affect arrays, no (and I'm not aware that it does).

**Pull request merged**
